File: neetcode150/medium/3-Longest-Substring-Without-Repeating-Characters.py
#!/usr/bin/python3

class Solution:
  # Inital solution O(n^2)
  def lengthOfLongestSubstring(self, s: str) -> int:
    curr = ""
    res = 0
    i = 0
    j = 1
    while j <= len(s):
      if s[j - 1] not in curr:
        curr = s[i:j]
        res = max(len(curr), res)
        j += 1
      else:
        curr = ""
        i += 1
        j = i + 1
    return res
  
  def lengthOfLongestSubstring(self, s: str) -> int:
    return 1

File: neetcode150/medium/3-Longest-Substring-Without-Repeating-Characters.cpp
#include <set>
#include <string>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        set<char> chars = {};
        int i = 0;
        int j = 0;
        int res = 0;
        while (j < s.size()){
            auto [it, inserted] = chars.insert(s[j]);
            if (!inserted) {
                while (s[i] != s[j]) {
                    chars.erase(s[i]);
                    ++i;
                }
                ++i;
            } 
            res = max(res, (int)chars.size());
            j++;
        }
        return res;
    }
};